### ISSUE TITLE
$translate on missing key does not return the key, but a failed promise

### ISSUE THREAD

**Issue by RichardBradley** [angular-translate/angular-translate#1516]
### Subject of the issue

The [documentation says](https://angular-translate.github.io/docs/#/guide/03_using-translate-service) "If [the translation id] doesn't exist it'll probably just return the translation id."

On that basis, I would expect a call like:

    $translate("MISSING_KEY").then(function(txt) { console.log(txt); })

to log "MISSING_KEY".

However, it does nothing at all, because the promise returned from $translate is completed with a failure.

The correct code would be:

    $translate("MISSING_KEY").then(function(txt) { console.log(txt); }, function(txt) { console.log(txt); })

as you can see in the "`it('should return translation id if translation doesn\'t exist'`" test at [translate.spec.js:276](https://github.com/angular-translate/angular-translate/blob/adeaa3e3a19f46cc4d40de8665f7e4492fce8475/test/unit/service/translate.spec.js#L276)

This means that the documentation is misleading and makes code that uses the service difficult to write correctly. 

For example, [this code in `ng-admin`](https://github.com/marmelab/ng-admin/blob/c9dc3efdba38f8953535f2b6da715693f97c19c1/src/javascripts/ng-admin/Crud/delete/DeleteController.js#L47) will return nothing at all if the "errorMessage" translation key is not found:

                $translate(errorMessage, {
                    status: error && error.status,
                    details: error && error.data && typeof error.data === 'object' ? JSON.stringify(error.data) : {}
                }).then(text => notification.log(text, { addnCls: 'humane-flatty-error' }));

... because the author did not anticipate that the promise returned by `$translate` could fail. (I will raise a separate bug with `ng-admin` to try to fix this.)

### Your environment
* version of angular-translate: 2.11.0
* version of angular: 1.4.8
* which browser and its version: n/a

### Steps to reproduce

Reproduced in the unit test linked above.

### Expected behaviour

The promise returned by `$translate("MISSING_KEY")` should complete with the value `"MISSING_KEY"`

### Actual behaviour

The promise returned by `$translate("MISSING_KEY")` does not complete, but fails with an error value of `"MISSING_KEY"`

**Comment by knalli** [comment]
Well.

Reading it literally, you are right: It will not return (implicitly read) the requested translation id on the success callback.

But: That's not promises are working and used to be used. A promise have at least two callbacks: A success and a failure. If the requested key has been resolved successfully, it will call the success method (with the translated value). If not, it will call the failure method (with the requested key). It does not make sense reporting a failure/miss as a success.

Using a non asynchronous method of request, you have only two options for reporting an error -- either raise an exception (the caller has to try/catch) or change the result. But we are asynchronous in this specific API, so we should use the possibilities correctly.

Although I don't see a problem with the documentation (that's how promises are working), I'm open for an optimized way of describing it correctly.

Feel free making a PR, we are always happy and welcome it! :)

**Comment by RichardBradley** [comment]
> That's not promises are working and used to be used. A promise have at least two callbacks: A success and a failure.

Well, you say that, but if you look in the docs there are 9 calls to "$translate" which use a "then" operator, 8 of which pass only one callback.

Only the example in "[Migration Guide](https://angular-translate.github.io/docs/#/guide/21_migration-guide)" uses an error callback.

I'll create a doc PR

**Comment by knalli** [comment]
@RichardBradley Yes, that is right. It should highlight the error callback. Thank you!

It is up to the developer whether the error callback is required or not. Calling `$translate.instant()` or the filter, you are never faced with this question, because that kind of information will be never available.

**Comment by RichardBradley** [comment]
Also, it is strange to reject the promise with the translation id, rather than the error reason. The [docs for Promise](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise) say:

> A pending promise can become either fulfilled with a value, or rejected with a reason (error)

The promise ought to be rejected with a failure reason like `Error("No matching translation key found")`, not with the translation id. The caller will already have the translation key to hand.

So this code from the "Migration Guide":

```
var stringFromService;
$translate('translationId').then(function(translation) {
  stringFromService = translation;
}, function(translationId){
  // Unable to translate given translationId
  stringFromService = translationId;
});
```

ought to look more like:

```
var stringFromService;
var translationId = "MY_TEXT";
$translate(translationId).then(function(translation) {
  stringFromService = translation;
}, function(error) {
  // Unable to translate given translationId; fall back to displaying the ID to the user:
  stringFromService = translationId;
});
```

**Comment by knalli** [comment]
Well, I think the (new) promise document by mdn is only slightly older than our release of 2.0 :) Said this, I want to say this is a style which we have introduced some time ago (finally released early 2014).

There are already some things perhaps everyone would change in a new system (see #995), but the api (v2) is fixed.

So, basically: Yes. Would be better. But not in v2 any more.

**Comment by reda-alaoui** [comment]
Hello,

It seems that this behaviour appeared between 2.4.1 and 2.11.0
We discovered it when updating angular translate library.

We have a process that call angular-translate for a bunch of keys for translation. We store all related promises in an array and then call `$q.all` on it.

Before, when one of the key was missing, the promise was resolved anyway with the key of each missing translation.
Now, the rejection of one promise implies the rejection of all promises.

Moreover, I think this is akward since the filter does not behave this way. It just returns the key when the translation is missing.

Is it possible to fallback to previous behaviour by introducing a setting for example?
